Mathlib/Data/Nat/Factorial/NatCast.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir, María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández
+-/
+
+import Mathlib.Algebra.Algebra.Defs
+import Mathlib.Algebra.CharP.Defs
+import Mathlib.Algebra.Order.Field.Rat
+import Mathlib.Data.Int.GCD
+import Mathlib.Data.Nat.Factorial.Basic
+import Mathlib.Data.Nat.Prime.Basic
+
+/-!
+# Invertibility of factorials
+
+This file contains lemmas providing sufficient conditions for the cast of `n!` to a commutative
+(semi)ring `A` to be a unit.
+
+-/
+
+namespace IsUnit
+
+open Nat
+section Semiring
+
+variable {A : Type*} [CommSemiring A]
```

Comment:
This is true without commutativity.  An ugly proof is
```lean
theorem natCast_factorial_of_le {n : ℕ} (hn_fac : IsUnit (n ! : A))
    {m : ℕ} (hmn : m ≤ n) : IsUnit (m ! : A) := by
  obtain ⟨k, rfl⟩ := exists_add_of_le hmn
  clear hmn
  induction k generalizing m with
  | zero => simpa using hn_fac
  | succ k ih =>
    rw [← add_assoc, add_right_comm] at hn_fac
    have := @ih (m + 1) hn_fac
    rw [Nat.factorial_succ, Nat.cast_mul, Nat.cast_commute _ _ |>.isUnit_mul_iff] at this
    exact this.2
```
though there might be a slicker trick by going through `Set.center`.